### PR TITLE
Tidy up generator trees

### LINKED ISSUE #8464: FAIL: resource/stack TestSerializePropertyValue
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

Something to investigate (also failing on master): either a problem with the test, or we have regressed on serialization ability.

https://github.com/pulumi/pulumi/runs/4257430169?check_suite_focus=true. 

```
=== FAIL: resource/stack TestSerializePropertyValue (190.46s)
Error:     deployment_test.go:444: [rapid] failed after 88 tests: (*T).FailNow() called
        To reproduce, specify -run="TestSerializePropertyValue" -rapid.failfile="TestSerializePropertyValue-20211118222509-76075.fail" (or -rapid.seed=1314149388578197325)
        Failed test output:
```

### Steps to reproduce

1.
2.
3.
4.

Expected: 
Actual:

### PR DESCRIPTION
**Pull request by t0yv0**
# Description

Fixes #8464 

I played with #8464 repro which reproduces on master but I could not get the Go system to print the generated value or the error.

I was also noticing how slow it was generating the first example.

I also noticed that it doesn't reproduce if depth is lowered to 5.

The repro is basically this:

```
git co f98d39b84bc0e3a78a13dc09f80730e39a297b10 -b repro
cd pkg/resource/stack
go test -test.v -run="TestSerializePropertyValue" -rapid.seed=1314149388578197325

=== RUN   TestSerializePropertyValue
    deployment_test.go:444: [rapid] failed after 0 tests: (*T).FailNow() called
        To reproduce, specify -run="TestSerializePropertyValue" -rapid.failfile="TestSerializePropertyValue-20211119164007-93661.fail" (or -rapid.seed=1314149388578197325)
        Failed test output:
--- FAIL: TestSerializePropertyValue (105.05s)
FAIL

```

It is frying CPU for 105s.

It is quite concerning that I cannot print the value out. I am still playing with that a bit.

On a slight tangent, I found a way to generate recursive trees in Rapid that shrinks to smaller trees correctly and does not waste as many generator objects. Introducing this mechanically makes the test complete in <2s and I can no longer reproduce the problem. It might be possible that the reason for this is also accidentally lowering maxDepth by 1, but I think we want to take the change anyway.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by t0yv0** [comment]
This modification still fails for me:

```
func TestSerializePropertyValue(t *testing.T) {
	gen := resource_testing.PropertyValueGenerator(6)
	rapid.Check(t, func(t *rapid.T) {
		gen.Draw(t, "property value")
	})
}
```

Basically it seems that `gen.Draw` is able to "quietly kill the test" and call `t.FailNow()` for some reason. This is not great, it is some issue in `rapid`. But at least it is reassuring that we do not have a failure of our property hiding in here.

**Comment by pgavlin** [comment]
> Basically it seems that gen.Draw is able to "quietly kill the test" and call t.FailNow() for some reason. This is not great, it is some issue in rapid. But at least it is reassuring that we do not have a failure of our property hiding in here.

Have you tried running this under a debugger and setting a breakpoint on `runtime.gopanic`?

**Comment by t0yv0** [comment]
Ha, well learning to use a debugger is still in my backlog..

**Comment by pgavlin** [comment]
I generally use [`dlv`](https://github.com/go-delve/delve) directly from a terminal after building the test binary in debug mode:

```bash
> go test -c -gcflags="all=-N -l" .
> dlv exec ./foo.test -- -test.run TheTestToRun
(dlv) b runtime.gopanic
(dlv) c
```

**Comment by t0yv0** [comment]
Ok yeah that worked. It points to LiteralArchiveGenerator using `require.NoError` where it probably shouldn't. Tinkering.

**Comment by t0yv0** [comment]
OK @pgavlin this is better: https://github.com/pulumi/pulumi/pull/8469

File: sdk/go/common/resource/testing/rapid.go [hunk 1]
```diff
@@ -351,64 +375,87 @@ func outputPropertyGenerator(ctx *StackContext, maxDepth int) *rapid.Generator {
...
+				arrayPropertyGenerator(self),
+				objectPropertyGenerator(self),
+				outputPropertyGenerator(ctx, self),
+				secretPropertyGenerator(self),
+			)
+		},
+	})
+}
 
-	if ctx == nil || len(ctx.Resources()) > 0 {
```
Comment (t0yv0) [inline sdk/go/common/resource/testing/rapid.go:401]: This should have stayed.

**Comment by t0yv0** [comment]
Played with it a bit more, it's curious .. actually no improvement on shrinking behavior - the old code is also left-biased via OneOf so that works out well; but the other surprising part was that this version is slower. Go might like the short-lived or stack-allocated generators better. I's interesting. Anyway, closing, not an improvement.

**Comment by t0yv0** [comment]
For future ref

```
go test -test.v -run="TestSerializePropertyValue" -rapid.seed=1314149388578197325
```

```
dlv test -- -test.v -run="TestSerializePropertyValue" -rapid.seed=1314149388578197325
```

**Pull request closed without merging**